src/native/managed/cdacreader/src/Contracts/RuntimeTypeSystem_1.cs

```diff
@@ -71,26 +73,77 @@ internal enum TypeHandleBits
         ValidMask = 2,
     }
 
+    internal enum MethodDescClassification
+    {
+        mcIL = 0, // IL
+        mcFCall = 1, // FCall (also includes tlbimped ctor, Delegate ctor)
+        mcNDirect = 2, // N/Direct
+        mcEEImpl = 3, // special method; implementation provided by EE (like Delegate Invoke)
+        mcArray = 4, // Array ECall
+        mcInstantiated = 5, // Instantiated generic methods, including descriptors
+                            // for both shared and unshared code (see InstantiatedMethodDesc)
+        mcComInterop = 6,
+        mcDynamic = 7, // for method desc with no metadata behind
+    }
+
     [Flags]
     internal enum MethodDescFlags : ushort
     {
+        ClassificationMask = 0x7,
         HasNonVtableSlot = 0x0008,
     }
 
+    internal enum InstantiatedMethodDescFlags2 : ushort
+    {
+        KindMask = 0x07,
+        GenericMethodDefinition = 0x01,
+        UnsharedMethodInstantiation = 0x02,
+        SharedMethodInstantiation = 0x03,
+        WrapperStubWithInstantiations = 0x04,
+    }
+
+    [Flags]
+    internal enum DynamicMethodDescExtendedFlags : uint
+    {
+        IsLCGMethod = 0x00004000,
+        IsILStub = 0x00008000,
+    }
+
     internal struct MethodDesc
     {
         private readonly Data.MethodDesc _desc;
         private readonly Data.MethodDescChunk _chunk;
+        private readonly Target _target;
+
         internal TargetPointer Address { get; init; }
-        internal MethodDesc(TargetPointer methodDescPointer, Data.MethodDesc desc, Data.MethodDescChunk chunk)
+        internal MethodDesc(Target target, TargetPointer methodDescPointer, Data.MethodDesc desc, Data.MethodDescChunk chunk)
         {
+            _target = target;
             _desc = desc;
             _chunk = chunk;
             Address = methodDescPointer;
         }
 
         public TargetPointer MethodTable => _chunk.MethodTable;
         public ushort Slot => _desc.Slot;
+        public uint Token
+        {
+            get
+            {
+                int tokenRemainderBitCount = _target.ReadGlobal<byte>(Constants.Globals.MethodDescTokenRemainderBitCount);
```

Comment:
Can we either store the token remainder bit count in the MethodDesc, or just pre-compose the token when creating the MethodDesc instance?  I've been avoiding storing `Target` and contracts in data structures, so they're just dumb value holders, not fancy algorithms.